scripts/mayachemtools/docs/modules/man3/Fingerprints.3
.\" Automatically generated by Pod::Man 2.28 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "Fingerprints 1"
.TH Fingerprints 1 "2018-05-15" "perl v5.22.4" "MayaChemTools"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
Fingerprints \- Fingerprints class
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
use Fingerprints::Fingerprints;
.PP
use Fingerprints::Fingerprints qw(:all);
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBFingerprints\fR class provides the following methods:
.PP
new, FoldFingerprintsByBitDensity, FoldFingerprintsBySize,
GetFingerprintBitsAsBinaryString, GetFingerprintBitsAsHexadecimalString,
GetFingerprintBitsAsRawBinaryString, GetFingerprintsVectorValueIDs,
GetFingerprintsVectorValues, IsFingerprintsGenerationSuccessful,
SetFingerprintsBitVector, SetFingerprintsVector, SetFingerprintsVectorType,
SetMolecule, SetSize, SetType, SetVectorType
.PP
\&\fBFingerprints\fR class is used as a base class for various specific fingerprint classes such as
\&\fBAtomNeighborhoodsFingerprints\fR, \fBAtomTypesFingerprints\fR, \fBEStateIndiciesFingerprints\fR,
\&\fBPathLengthFingerprints\fR, \fBExtendedConnectivityFingerprints\fR, \fBMACCSKeys\fR and so on.
It implements functionality common to fingerprint classes.
.PP
\&\fBFingerprints\fR class is  derived from \fBObjectProperty\fR base class which provides methods not
explicitly defined in \fBFingerprints\fR or \fBObjectProperty\fR classes using Perl's \s-1AUTOLOAD\s0 functionality.
These methods are generated on-the-fly for a specified object property:
.PP
.Vb 3
\&    Set<PropertyName>(<PropertyValue>);
\&    $PropertyValue = Get<PropertyName>();
\&    Delete<PropertyName>();
.Ve
.PP
\&\fBFingerprints\fR class uses \fBFingerprintsBitVector\fR class to provide bits manipulation functionality.
.SS "\s-1METHODS\s0"
.IX Subsection "METHODS"
.IP "\fBnew\fR" 4
.IX Item "new"
.Vb 1
\&    $NewFingerprints = new Fingerprints(%NamesAndValues);
.Ve
.Sp
Using specified \fIFingerprints\fR property names and values hash, \fBnew\fR method creates a new object
and returns a reference to newly created \fBFingerprints\fR object. By default, following properties are
initialized:
.Sp
.Vb 9
\&    Molecule = \*(Aq\*(Aq;
\&    Type = \*(Aq\*(Aq;
\&    VectorType = \*(Aq\*(Aq;
\&    Size = \*(Aq\*(Aq;
\&    MinSize = \*(Aq\*(Aq;
\&    MaxSize = \*(Aq\*(Aq;
\&    FingerprintsBitVector = \*(Aq\*(Aq;
\&    FingerprintsVectorType = \*(Aq\*(Aq;
\&    FingerprintsVector = \*(Aq\*(Aq;
.Ve
.IP "\fBFoldFingerprintsByBitDensity\fR" 4
.IX Item "FoldFingerprintsByBitDensity"
.Vb 1
\&    $Fingerprints\->FoldFingerprintsByBitDensity($BitDensity);
.Ve
.Sp
Folds fingerprints by recursively reducing its size by half until bit density is greater than or equal to
specified \fIBitDensity\fR and returns \fIFingerprints\fR.
.IP "\fBFoldFingerprintsBySize\fR" 4
.IX Item "FoldFingerprintsBySize"
.Vb 1
\&    $Fingerprints\->FoldFingerprintsBySize($Size, [$CheckSizeValue]);
.Ve
.Sp
Fold fingerprints by recursively reducing its size by half until size is less than or equal to specified
\&\fISize\fR and returns \fIFingerprints\fR. By default, value \fISize\fR is checked to make sure it's:
.Sp
.Vb 1
\&    >= MinSize and < Size and IsPowerOfTwo
.Ve
.IP "\fBGetFingerprintBitsAsBinaryString\fR" 4
.IX Item "GetFingerprintBitsAsBinaryString"
.Vb 2
\&    $BinaryASCIIString =
\&       $Fingerprints\->GetFingerprintBitsAsBinaryString();
.Ve
.Sp
Returns fingerprints as a binary \s-1ASCII\s0 string containing 0s and 1s.
.IP "\fBGetFingerprintBitsAsHexadecimalString\fR" 4
.IX Item "GetFingerprintBitsAsHexadecimalString"
.Vb 2
\&    $HexadecimalString =
\&       $Fingerprints\->GetFingerprintBitsAsHexadecimalString();
.Ve
.Sp
Returns fingerprints as a hexadecimal string
.IP "\fBGetFingerprintBitsAsRawBinaryString\fR" 4
.IX Item "GetFingerprintBitsAsRawBinaryString"
.Vb 2
\&    $RawBinaryString =
\&       $Fingerprints\->GetFingerprintBitsAsRawBinaryString();
.Ve
.Sp
Returns fingerprints as a raw binary string containing packed bit values for each byte.
.IP "\fBGetFingerprintsVectorValueIDs\fR" 4
.IX Item "GetFingerprintsVectorValueIDs"
.Vb 2
\&    $ValueIDsRef = $Fingerprints\->GetFingerprintsVectorValueIDs();
\&    @ValueIDs = $Fingerprints\->GetFingerprintsVectorValueIDs();
.Ve
.Sp
Returns fingerprints vector value IDs as an array or reference to an array.
.IP "\fBGetFingerprintsVectorValues\fR" 4
.IX Item "GetFingerprintsVectorValues"
.Vb 2
\&    $ValuesRef = $Fingerprints\->GetFingerprintsVectorValues();
\&    @Values = $Fingerprints\->GetFingerprintsVectorValues();
.Ve
.Sp
Returns fingerprints vector values as an array or reference to an array.
.IP "\fBIsFingerprintsGenerationSuccessful\fR" 4
.IX Item "IsFingerprintsGenerationSuccessful"
.Vb 1
\&    $Return = $Fingerprints\->IsFingerprintsGenerationSuccessful();
.Ve
.Sp
Returns 1 or 0 based on whether fingerprints were successfully generated.
.IP "\fBSetFingerprintsBitVector\fR" 4
.IX Item "SetFingerprintsBitVector"
.Vb 1
\&    $Fingerprints\->SetFingerprintsBitVector($FingerprintsBitVector);
.Ve
.Sp
Sets \fIFingerprintsBitVector\fR object for \fIFingerprints\fR and returns \fIFingerprints\fR.
.IP "\fBSetFingerprintsVector\fR" 4
.IX Item "SetFingerprintsVector"
.Vb 1
\&    $Fingerprints\->SetFingerprintsVector();
.Ve
.Sp
Sets \fIFingerprintsVector\fR object for \fIFingerprints\fR and returns \fIFingerprints\fR.
.IP "\fBSetFingerprintsVectorType\fR" 4
.IX Item "SetFingerprintsVectorType"
.Vb 1
\&    $Fingerprints\->SetFingerprintsVectorType($VectorType);
.Ve
.Sp
Sets \fIFingerprintsVector\fR type for \fIFingerprints\fR and returns \fIFingerprints\fR. Possible
\&\fIVectorType\fR values: \fIOrderedNumericalValues, NumericalValues or AlphaNumericalValues\fR.
.IP "\fBSetMolecule\fR" 4
.IX Item "SetMolecule"
.Vb 1
\&    $Fingerprints\->SetMolecule($Molecule);
.Ve
.Sp
Sets \fIMolecule\fR object for \fIFingerprints\fR and returns \fIFingerprints\fR.
.IP "\fBSetSize\fR" 4
.IX Item "SetSize"
.Vb 1
\&    $Fingerprints\->SetSize($Size);
.Ve
.Sp
Sets \fISize\fR of fingerprints and returns \fIFingerprints\fR.
.IP "\fBSetType\fR" 4
.IX Item "SetType"
.Vb 1
\&    $Fingerprints\->SetType($Type);
.Ve
.Sp
Sets \fIType\fR of fingerprints and returns \fIFingerprints\fR.
.IP "\fBSetVectorType\fR" 4
.IX Item "SetVectorType"
.Vb 1
\&    $Fingerprints\->SetVectorType($Type);
.Ve
.Sp
Sets \fIType\fR of fingerprints vector and returns \fIFingerprints\fR. Possible \fIType\fR values:
\&\fIFingerprintsBitVector or FingerprintsVector\fR.
.SH "AUTHOR"
.IX Header "AUTHOR"
Manish Sud <msud@san.rr.com>
.SH "SEE ALSO"
.IX Header "SEE ALSO"
FingerprintsStringUtil.pm, AtomNeighborhoodsFingerprints.pm, AtomTypesFingerprints.pm,
EStateIndiciesFingerprints.pm, ExtendedConnectivityFingerprints.pm, MACCSKeys.pm,
PathLengthFingerprints.pm, TopologicalAtomPairsFingerprints.pm, TopologicalAtomTripletsFingerprints.pm,
TopologicalAtomTorsionsFingerprints.pm, TopologicalPharmacophoreAtomPairsFingerprints.pm,
TopologicalPharmacophoreAtomTripletsFingerprints.pm
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright (C) 2018 Manish Sud. All rights reserved.
.PP
This file is part of MayaChemTools.
.PP
MayaChemTools is free software; you can redistribute it and/or modify it under
the terms of the \s-1GNU\s0 Lesser General Public License as published by the Free
Software Foundation; either version 3 of the License, or (at your option)
any later version.


mayachemtools/lib/Fingerprints/Fingerprints.pm
package Fingerprints::Fingerprints;
#
# File: Fingerprints.pm
# Author: Manish Sud <msud@san.rr.com>
#
# Copyright (C) 2018 Manish Sud. All rights reserved.
#
# This file is part of MayaChemTools.
#
# MayaChemTools is free software; you can redistribute it and/or modify it under
# the terms of the GNU Lesser General Public License as published by the Free
# Software Foundation; either version 3 of the License, or (at your option) any
# later version.
#
# MayaChemTools is distributed in the hope that it will be useful, but without
# any warranty; without even the implied warranty of merchantability of fitness
# for a particular purpose.  See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with MayaChemTools; if not, see <http://www.gnu.org/licenses/> or
# write to the Free Software Foundation Inc., 59 Temple Place, Suite 330,
# Boston, MA, 02111-1307, USA.
#

use strict;
use Carp;
use Exporter;
use Scalar::Util ();
use ObjectProperty;
use MathUtil ();
use TextUtil ();
use Fingerprints::FingerprintsBitVector;
use Fingerprints::FingerprintsVector;

use vars qw(@ISA @EXPORT @EXPORT_OK %EXPORT_TAGS);

@ISA = qw(ObjectProperty Exporter);
@EXPORT = qw();
@EXPORT_OK = qw();

%EXPORT_TAGS = (all  => [@EXPORT, @EXPORT_OK]);

# Setup class variables...
my($ClassName);
_InitializeClass();

# Class constructor...
sub new {
  my($Class, %NamesAndValues) = @_;

  # Initialize object...
  my $This = {};
  bless $This, ref($Class) || $Class;
  $This->_InitializeFingerprints();

  $This->_InitializeFingerprintsProperties(%NamesAndValues);

  return $This;
}

# Initialize object data...
#
sub _InitializeFingerprints {
  my($This) = @_;

  # Molecule object...
  $This->{Molecule} = '';

  # Type of fingerprints...
  $This->{Type} = '';

  # Type of fingerprints vector: FingerprintsBitVector or FingerprintsVector...
  $This->{VectorType} = '';

  # Marks successful generation of fingerprints...
  $This->{FingerprintsGenerated} = 0;

  # Initialize values for FingerprintsBitVector...
  _InitializeClassValuesForFingerprintsBitVector();

  # Initialize values for FingerprintsVector...
  _InitializeClassValuesForFingerprintsVector();
}

# Initialize class ...
sub _InitializeClass {
  #Class name...
  $ClassName = __PACKAGE__;
}

# Initialize class values specific to FingerprintsBitVector...
#
sub _InitializeClassValuesForFingerprintsBitVector {
  my($This) = @_;

  # Size of FingerprintsBitVector...
  $This->{Size} = '';

  # Min/Max sizes used for folding FingerprintsBitVector...
  $This->{MinSize} = '';
  $This->{MaxSize} = '';

  # FingerprintsBitVector...
  $This->{FingerprintsBitVector} = '';
}

# Initialize class values specific to FingerprintsVector...
#
sub _InitializeClassValuesForFingerprintsVector {
  my($This) = @_;

  # Types of FingerprintsVector values: OrderedNumericalValues, NumericalValues or AlphaNumericalValues.
  $This->{FingerprintsVectorType} = '';

  # Fingerprints vector...
  $This->{FingerprintsVector} = '';
}

# Initialize object properties....
sub _InitializeFingerprintsProperties {
  my($This, %NamesAndValues) = @_;

  my($Name, $Value, $MethodName);
  while (($Name, $Value) = each  %NamesAndValues) {
    $MethodName = "Set${Name}";
    $This->$MethodName($Value);
  }

  return $This;
}

# Set molecule object and make sure it's not already set...
#
sub SetMolecule {
  my($This, $Molecule) = @_;

  if ($This->{Molecule}) {
    croak "Error: ${ClassName}->SetMolecule: Can't change molecule object:  It's already set...";
  }
  $This->{Molecule} = $Molecule;

  # Weaken the reference to disable increment of reference count...
  Scalar::Util::weaken($This->{Molecule});

  return $This;
}

# Set type and make sure it's not already set...
#
sub SetType {
  my($This, $Type) = @_;

  if ($This->{Type}) {
    croak "Error: ${ClassName}->SetType: Can't change fingerprint type:  It's already set...";
  }
  $This->{Type} = $Type;

  return $This;
}

# Is fingerprints generation successful?
#
# Notes:
#   . The specific fingerprints generation class sets the value of FingerprinsGenerated
#     to 1 after the successful generation of fingerprints; otherwise, it's set to 0.
#
sub IsFingerprintsGenerationSuccessful {
  my($This) = @_;

  return $This->{FingerprintsGenerated} ? 1 : 0;
}

# Set vector type and make sure it's not already set...
#
sub SetVectorType {
  my($This, $VectorType) = @_;

  if ($This->{VectorType}) {
    croak "Error: ${ClassName}->SetVectorType: Can't change fingerprint vector type:  It's already set...";
  }
  if ($VectorType !~ /^(FingerprintsBitVector|FingerprintsVector)$/i) {
    croak "Error: ${ClassName}->SetVectorType: Specified value, $VectorType, for Type is not vaild. Supported types in current release of MayaChemTools: FingerprintsBitVector and FingerprintsVector...";
  }
  $This->{VectorType} = $VectorType;

  return $This;
}

#
# Methods/functions for FingerprintsBitVector...
#

# Initialize fingerprint bit vector...
sub _InitializeFingerprintsBitVector {
  my($This) = @_;

  if ($This->{VectorType} !~ /^FingerprintsBitVector$/i) {
    croak "Error: ${ClassName}->_InitializeFingerprintsBitVector: Can't initialize fingerprints bit vector: VectorType must be FingerprintsBitVector...";
  }

  if ($This->{Size}) {
    $This->{FingerprintsBitVector} = new Fingerprints::FingerprintsBitVector($This->{Size});
  }
  return $This;
}

# Set size...
#
sub SetSize {
  my($This, $Size) = @_;

  if ($This->{MinSize} && $Size < $This->{MinSize}) {
    croak "Error: ${ClassName}->SetSize: Fingerprint size value, $Size, is not valid :  It must be >= $This->{MinSize}...";
  }
  if ($This->{MaxSize} && $Size > $This->{MaxSize}) {
    croak "Error: ${ClassName}->SetSize: Fingerprint size value, $Size, is not valid:  It must be <= $This->{MaxSize}...";
  }

  $This->{Size} = $Size;

  return $This;
}

# Set FingerprintsBitVector object and make sure it's not already set...
#
sub SetFingerprintsBitVector {
  my($This, $FingerprintsBitVector) = @_;

  if ($This->{FingerprintsBitVector}) {
    croak "Error: ${ClassName}->SetFingerprintsBitVector: Can't change FingerprintsBitVector object:  It's already set...";
  }
  $This->{FingerprintsBitVector} = $FingerprintsBitVector;

  return $This;
}

# Fold fingerprints by recursively reducing its size by half untill bit density is greater than or equal to
# specified bit density...
#
sub FoldFingerprintsByBitDensity {
  my($This, $BitDensity) = @_;

  if (!($BitDensity >= 0 && $BitDensity <= 1)) {
    croak "Error: ${ClassName}->FoldFingerprintsByBitDensity: Specified bit density, $BitDensity, is not valid:  It must be > 0 && <= 1 ...";
  }

  return $This->_FoldFingerprintsBitVector('ByDensity', $BitDensity);
}

# Fold fingerprints by recursively reducing its size by half untill size is less than or equal to
# specified size...
#
sub FoldFingerprintsBySize {
  my($This, $Size, $CheckSizeValue) = @_;

  if (!defined $CheckSizeValue) {
    $CheckSizeValue = 1;
  }

  if ($CheckSizeValue) {
    if (!TextUtil::IsPositiveInteger($Size)) {
      croak "Error: ${ClassName}->FoldFingerprintsBySize: Specified size, $Size, is not valid:  It must be a positive integer";
    }
    if (!($Size >= $This->{MinSize} && $Size < $This->{Size})) {
      croak "Error: ${ClassName}->FoldFingerprintsBySize: Specified size, $Size, is not valid:  It must be greater than or equal to minimum size of $This->{MinSize} and less than current size of $This->{Size}...";
    }
    if (!TextUtil::IsNumberPowerOfNumber($Size, 2)) {
      croak "Error: ${ClassName}->FoldFingerprintsBySize: Specified size value, $Size, must be power of 2...";
    }
  }

  return $This->_FoldFingerprintsBitVector('BySize', $Size);
}

# Fold fingerprints bit vector using specified size of bit density...
#
sub _FoldFingerprintsBitVector {
  my($This, $Mode, $Value) = @_;

  if (!$This->{FingerprintsBitVector}) {
    return $This;
  }
  my($FingerprintsBitVector, $FoldedFingerprintsBitVector);

  $FoldedFingerprintsBitVector = '';
  $FingerprintsBitVector = $This->{FingerprintsBitVector};
  MODE: {
    if ($Mode =~ /^BySize$/i) { $FoldedFingerprintsBitVector = $FingerprintsBitVector->FoldFingerprintsBitVectorBySize($Value); last MODE; }
    if ($Mode =~ /^ByDensity$/i) { $FoldedFingerprintsBitVector = $FingerprintsBitVector->FoldFingerprintsBitVectorByDensity($Value); last MODE; }
    $FoldedFingerprintsBitVector = '';
  }
  if ($FoldedFingerprintsBitVector) {
    $This->{FingerprintsBitVector} = $FoldedFingerprintsBitVector;
    $This->{Size} = $FoldedFingerprintsBitVector->GetSize();
  }
  return $This;
}

# Get fingerprints as a binary ascii string containing 0s and 1s...
#
sub GetFingerprintBitsAsBinaryString {
  my($This, $BitOrder) = @_;

  return $This->_GetFingerprintBitsAsString('Binary', $BitOrder);
}

# Get fingerprints as a hexadecimal string...
#
sub GetFingerprintBitsAsHexadecimalString {
  my($This, $BitOrder) = @_;

  return $This->_GetFingerprintBitsAsString('Hexadecimal', $BitOrder);
}

# Get fingerprints as a raw binary string containing packed bit values for each
# byte...
#
sub GetFingerprintBitsAsRawBinaryString {
  my($This, $BitOrder) = @_;

  return $This->_GetFingerprintBitsAsString('RawBinary', $BitOrder);
}

# Get fingerprint bits as a string...
#
sub _GetFingerprintBitsAsString {
  my($This, $Format) = @_;

  if (!$This->{FingerprintsBitVector}) {
    return undef;
  }
  FORMAT : {
    if ($Format =~ /^(Binary|Bin|BinaryString)$/i) { return $This->{FingerprintsBitVector}->GetBitsAsBinaryString(); last FORMAT; }
    if ($Format =~ /^(Hexadecimal|Hex|HexadecimalString)$/i) { return $This->{FingerprintsBitVector}->GetBitsAsHexadecimalString(); last FORMAT; }
    if ($Format =~ /^(RawBinary|RawBin|RawBinaryString)$/i) { return $This->{FingerprintsBitVector}->GetBitsAsRawBinaryString(); last FORMAT; }
    croak "Error: ${ClassName}->_GetFingerprintBitsAsString: Specified bit vector string format, $Format, is not supported. Value values: Binary, Bin, BinaryString, Hexdecimal, Hex, HexadecimalString, RawBinary, RawBin, RawBinaryString...";
  }
  return undef;
}

#
# Methods/functions for FingerprintsVector...
#

# Initialize fingerprint vector...
sub _InitializeFingerprintsVector {
  my($This) = @_;

  if ($This->{VectorType} !~ /^FingerprintsVector$/i) {
    croak "Error: ${ClassName}->_InitializeFingerprintsVector: Can't initialize fingerprints vector: VectorType must be FingerprintsVector...";
  }

  if ($This->{FingerprintsVectorType}) {
    $This->{FingerprintsVector} = new Fingerprints::FingerprintsVector('Type' => $This->{FingerprintsVectorType});
  }
  return $This;
}

# Set FingerprintsVector object and make sure it's not already set...
#
sub SetFingerprintsVector {
  my($This, $FingerprintsVector) = @_;

  if ($This->{FingerprintsVector}) {
    croak "Error: ${ClassName}->SetFingerprintsVector: Can't change FingerprintsVector object:  It's already set...";
  }
  $This->{FingerprintsVector} = $FingerprintsVector;

  return $This;
}

# Types of FingerprintsVector values: OrderedNumericalValues, NumericalValues or AlphaNumericalValues.
#
sub SetFingerprintsVectorType {
  my($This, $FingerprintsVectorType) = @_;

  if ($This->{FingerprintsVectorType}) {
    croak "Error: ${ClassName}->SetFingerprintsVector: Can't change FingerprintsVectorType:  It's already set...";
  }
  if ($FingerprintsVectorType !~ /^(OrderedNumericalValues|NumericalValues|AlphaNumericalValues)$/i) {
    croak "Error: ${ClassName}->SetFingerprintsVectorType: Specified value, $FingerprintsVectorType, for Type is not vaild. Supported types in current release of MayaChemTools: OrderedNumericalValues, NumericalValues or AlphaNumericalValues";
  }
  $This->{FingerprintsVectorType} = $FingerprintsVectorType;

  return $This;
}

# Get fingerprints vector values as an array or reference to an array...
#
sub GetFingerprintsVectorValues {
  my($This) = @_;

  if (!$This->{FingerprintsVector}) {
    return undef;
  }
  return $This->{FingerprintsVector}->GetValues();
}

# Get fingerprints vector value IDs as an array or reference to an array...
#
sub GetFingerprintsVectorValueIDs {
  my($This) = @_;

  if (!$This->{FingerprintsVector}) {
    return undef;
  }
  return $This->{FingerprintsVector}->GetValueIDs();
}

1;

__END__

=head1 NAME

Fingerprints - Fingerprints class

=head1 SYNOPSIS

use Fingerprints::Fingerprints;

use Fingerprints::Fingerprints qw(:all);

=head1 DESCRIPTION

B<Fingerprints> class provides the following methods:

new, FoldFingerprintsByBitDensity, FoldFingerprintsBySize,
GetFingerprintBitsAsBinaryString, GetFingerprintBitsAsHexadecimalString,
GetFingerprintBitsAsRawBinaryString, GetFingerprintsVectorValueIDs,
GetFingerprintsVectorValues, IsFingerprintsGenerationSuccessful,
SetFingerprintsBitVector, SetFingerprintsVector, SetFingerprintsVectorType,
SetMolecule, SetSize, SetType, SetVectorType

B<Fingerprints> class is used as a base class for various specific fingerprint classes such as
B<AtomNeighborhoodsFingerprints>, B<AtomTypesFingerprints>, B<EStateIndiciesFingerprints>,
B<PathLengthFingerprints>, B<ExtendedConnectivityFingerprints>, B<MACCSKeys> and so on.
It implements functionality common to fingerprint classes.

B<Fingerprints> class is  derived from B<ObjectProperty> base class which provides methods not
explicitly defined in B<Fingerprints> or B<ObjectProperty> classes using Perl's AUTOLOAD functionality.
These methods are generated on-the-fly for a specified object property:

    Set<PropertyName>(<PropertyValue>);
    $PropertyValue = Get<PropertyName>();
    Delete<PropertyName>();

B<Fingerprints> class uses B<FingerprintsBitVector> class to provide bits manipulation functionality.

=head2 METHODS

=over 4

=item B<new>

    $NewFingerprints = new Fingerprints(%NamesAndValues);

Using specified I<Fingerprints> property names and values hash, B<new> method creates a new object
and returns a reference to newly created B<Fingerprints> object. By default, following properties are
initialized:

    Molecule = '';
    Type = '';
    VectorType = '';
    Size = '';
    MinSize = '';
    MaxSize = '';
    FingerprintsBitVector = '';
    FingerprintsVectorType = '';
    FingerprintsVector = '';

=item B<FoldFingerprintsByBitDensity>

    $Fingerprints->FoldFingerprintsByBitDensity($BitDensity);

Folds fingerprints by recursively reducing its size by half until bit density is greater than or equal to
specified I<BitDensity> and returns I<Fingerprints>.

=item B<FoldFingerprintsBySize>

    $Fingerprints->FoldFingerprintsBySize($Size, [$CheckSizeValue]);

Fold fingerprints by recursively reducing its size by half until size is less than or equal to specified
I<Size> and returns I<Fingerprints>. By default, value I<Size> is checked to make sure it's:

    >= MinSize and < Size and IsPowerOfTwo

=item B<GetFingerprintBitsAsBinaryString>

    $BinaryASCIIString =
       $Fingerprints->GetFingerprintBitsAsBinaryString();

Returns fingerprints as a binary ASCII string containing 0s and 1s.

=item B<GetFingerprintBitsAsHexadecimalString>

    $HexadecimalString =
       $Fingerprints->GetFingerprintBitsAsHexadecimalString();

Returns fingerprints as a hexadecimal string

=item B<GetFingerprintBitsAsRawBinaryString>

    $RawBinaryString =
       $Fingerprints->GetFingerprintBitsAsRawBinaryString();

Returns fingerprints as a raw binary string containing packed bit values for each byte.

=item B<GetFingerprintsVectorValueIDs>

    $ValueIDsRef = $Fingerprints->GetFingerprintsVectorValueIDs();
    @ValueIDs = $Fingerprints->GetFingerprintsVectorValueIDs();

Returns fingerprints vector value IDs as an array or reference to an array.

=item B<GetFingerprintsVectorValues>

    $ValuesRef = $Fingerprints->GetFingerprintsVectorValues();
    @Values = $Fingerprints->GetFingerprintsVectorValues();

Returns fingerprints vector values as an array or reference to an array.

=item B<IsFingerprintsGenerationSuccessful>

    $Return = $Fingerprints->IsFingerprintsGenerationSuccessful();

Returns 1 or 0 based on whether fingerprints were successfully generated.

=item B<SetFingerprintsBitVector>

    $Fingerprints->SetFingerprintsBitVector($FingerprintsBitVector);

Sets I<FingerprintsBitVector> object for I<Fingerprints> and returns I<Fingerprints>.

=item B<SetFingerprintsVector>

    $Fingerprints->SetFingerprintsVector();

Sets I<FingerprintsVector> object for I<Fingerprints> and returns I<Fingerprints>.

=item B<SetFingerprintsVectorType>

    $Fingerprints->SetFingerprintsVectorType($VectorType);

Sets I<FingerprintsVector> type for I<Fingerprints> and returns I<Fingerprints>. Possible
I<VectorType> values: I<OrderedNumericalValues, NumericalValues or AlphaNumericalValues>.

=item B<SetMolecule>

    $Fingerprints->SetMolecule($Molecule);

Sets I<Molecule> object for I<Fingerprints> and returns I<Fingerprints>.

=item B<SetSize>

    $Fingerprints->SetSize($Size);

Sets I<Size> of fingerprints and returns I<Fingerprints>.

=item B<SetType>

    $Fingerprints->SetType($Type);

Sets I<Type> of fingerprints and returns I<Fingerprints>.

=item B<SetVectorType>

    $Fingerprints->SetVectorType($Type);

Sets I<Type> of fingerprints vector and returns I<Fingerprints>. Possible I<Type> values:
I<FingerprintsBitVector or FingerprintsVector>.

=back

=head1 AUTHOR

Manish Sud <msud@san.rr.com>

=head1 SEE ALSO

FingerprintsStringUtil.pm, AtomNeighborhoodsFingerprints.pm, AtomTypesFingerprints.pm,
EStateIndiciesFingerprints.pm, ExtendedConnectivityFingerprints.pm, MACCSKeys.pm,
PathLengthFingerprints.pm, TopologicalAtomPairsFingerprints.pm, TopologicalAtomTripletsFingerprints.pm,
TopologicalAtomTorsionsFingerprints.pm, TopologicalPharmacophoreAtomPairsFingerprints.pm,
TopologicalPharmacophoreAtomTripletsFingerprints.pm

=head1 COPYRIGHT

Copyright (C) 2018 Manish Sud. All rights reserved.

This file is part of MayaChemTools.

MayaChemTools is free software; you can redistribute it and/or modify it under
the terms of the GNU Lesser General Public License as published by the Free
Software Foundation; either version 3 of the License, or (at your option)
any later version.

=cut
